### PR TITLE
[RFC][flang] Replace special symbols in uniqued global names.

### PR DESCRIPTION
**Pull request by vzakhari**
This change addresses more "issues" as the one resolved in #71338.
Some targets (e.g. NVPTX) do not accept global names containing
`.`. In particular, the global variables created to represent
the runtime information of derived types use `.` in their names.
A derived type's descriptor object may be used in the device code,
e.g. to initialize a descriptor of a variable of this type.
Thus, the runtime type info objects may need to be compiled
for the device.

Moreover, at least the derived types' descriptor objects
may need to be registered (think of `omp declare target`)
for the host-device association so that the addendum pointer
can be properly mapped to the device for descriptors using
a derived type's descriptor as their addendum pointer.
The registration implies knowing the name of the global variable
in the device image so that proper host code can be created.
So it is better to name the globals the same way for the host
and the device.

CompilerGeneratedNamesConversion pass renames all uniqued globals
such that the special symbols (currently `.`) are replaced
with `X`. The pass is supposed to be run for the host and the device.

An option is added to FIR-to-LLVM conversion pass to indicate
whether the new pass has been run before or not. This setting
affects how the codegen computes the names of the derived types'
descriptors for FIR derived types.

fir::NameUniquer now allows `X` to be part of a name, because
the name deconstruction may be applied to the mangled names
after CompilerGeneratedNamesConversion pass.

### CODE REVIEW

File: flang/lib/Optimizer/Transforms/CompilerGeneratedNames.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  auto *context = &getContext();
+
+  llvm::DenseMap<mlir::StringAttr, mlir::FlatSymbolRefAttr> remappings;
+  for (auto &funcOrGlobal : op->getRegion(0).front()) {
+    if (llvm::isa<mlir::func::FuncOp>(funcOrGlobal) ||
+        llvm::isa<fir::GlobalOp>(funcOrGlobal)) {
+      auto symName = funcOrGlobal.getAttrOfType<mlir::StringAttr>(
+          mlir::SymbolTable::getSymbolAttrName());
+      auto deconstructedName = fir::NameUniquer::deconstruct(symName);
+      if (deconstructedName.first != fir::NameUniquer::NameKind::NOT_UNIQUED &&
```
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/CompilerGeneratedNames.cpp:51]: IO lowering is producing some global symbols with dots for the purpose of type bound IO that may not be considered unique here I think.

See:
https://github.com/llvm/llvm-project/blob/34e15adb5a725a5ecc7c0d5ef5571d307d751a93/flang/test/Lower/io-derived-type.f90#L134
https://github.com/llvm/llvm-project/blob/34e15adb5a725a5ecc7c0d5ef5571d307d751a93/flang/lib/Lower/IO.cpp#L302

**Review by jeanPerier — APPROVED** [review]
Code looks good to me (small nit about an IO case probably not very relevant to your use case). I am wondering a bit if we should just use X right away as a separator rather than use a rewrite pass to avoid the pass complexity, but "." is more readable in the IR and the flexibility makes some sense.

File: flang/lib/Optimizer/Transforms/CompilerGeneratedNames.cpp (hunk 1 above)
Comment (vdonaldson) [reply flang/lib/Optimizer/Transforms/CompilerGeneratedNames.cpp:51]: In addition to the `.nonTbpDefinedIoTable` suffix, IO.cpp also has two instances of a `.list` suffix.

**Comment by vzakhari** [comment]
> Code looks good to me (small nit about an IO case probably not very relevant to your use case). I am wondering a bit if we should just use X right away as a separator rather than use a rewrite pass to avoid the pass complexity, but "." is more readable in the IR and the flexibility makes some sense.

Thank you for the review, Jean!  I tried using `X` earlier in the compilation pipeline, first, but then I failed to cleanly resolve the lowercasing that is done for the symbol names.  So I decided to follow your reasoning in https://github.com/llvm/llvm-project/pull/71338#issuecomment-1795750355 and put this into a pass.  I tried combining the logic with the `ExternalNamesConversion` pass, but that did not look clean to me and I ended up separating it into a new pass.

File: flang/lib/Optimizer/Transforms/CompilerGeneratedNames.cpp (hunk 1 above)
Comment (vzakhari) [reply flang/lib/Optimizer/Transforms/CompilerGeneratedNames.cpp:51]: Thank you for the pointers, Jean and Val!  I would like to address these cases separately.

These names do not look quite from the reserved name space to me, so I guess they may conflict with the user space if I try to replace `.` with `X` in them (e.g. when linking with an object file created from a C module).  I think I can try to put them into the "Flang's name space", e.g. `@_QEdefault.nonTbpDefinedIoTable` or `@_QECdefault.nonTbpDefinedIoTable`, then the new pass should work for them and we will be safe from the name conflicts (given that `_Q` is reserved).

Comment (jeanPerier) [reply flang/lib/Optimizer/Transforms/CompilerGeneratedNames.cpp:51]: Solution you suggest makes sense to me, I am OK doing it in separate patch.

**Comment by jeanPerier** [comment]
> Thank you for the review, Jean! I tried using `X` earlier in the compilation pipeline, first, but then I failed to cleanly resolve the lowercasing that is done for the symbol names. So I decided to follow your reasoning in [#71338 (comment)](https://github.com/llvm/llvm-project/pull/71338#issuecomment-1795750355) and put this into a pass. I tried combining the logic with the `ExternalNamesConversion` pass, but that did not look clean to me and I ended up separating it into a new pass.

Thanks for the link and explanation. You approach looks good to me.

**Comment by abidh** [comment]
Hi @vzakhari, @jeanPerier 
We discard compiler generated symbol (used for derived types) while generating debug info. Previously I could just look at the '.' in the start to be sure that I can discard this without worrying that it can be a symbol from user code. Is there a way now to do this after '.' has been replaced with 'X'. For example, `_QMmEXcXt1` is a compiler generated one and `_QMmEsome_global` is a user symbol and I don't see a straight forward way to differentiate between them. 

Without this, we will be generating a lot of redundant information which not only takes compile time but will increase the object size unnecessarily.

**Comment by vzakhari** [comment]
Hi @abidh, how does the '.' lookup work for `_QQclX9a37c0` symbols after https://github.com/llvm/llvm-project/pull/71338?  Is it also "broken"?

I think you can try to rely on 'X' symbol being the first in the uniqued decomposed name, but maybe we should actually have an attribute for the compiler-generated globals.

Please note that the not-uniqued symbols can have 'X' in their names, and that does not mean they are compiler-generated:
```
module test
  real, bind(c,name='XYZ') :: aaa
end module test
```

**Comment by abidh** [comment]
> real, bind(c,name='XYZ') :: aaa

For variable like `_QQclX9a37c0`, the kind is `NameKind::GENERATED` which we can use to filter them. The variable like `_QMmEXcXt1` have `NameKind::VARIABLE` so they go through. I was thinking that if we could get `NameKind::GENERATED` for all compiler generated variables then that probably would be the cleanest solution.

As a temporary solution, I could filter XcX* and XdtX* like variable by looking at their type which will have _QM__fortran_type_info*. But variable *XnX* will still be there as those have normal character types.

**Comment by vzakhari** [comment]
Right, it seems that using `NameKind::GENERATED` may be the right approach.

**Pull request merged**